### ISSUE TITLE
Should we add registerSubclass?

### ISSUE THREAD

**Issue by maaeps** [parse-community/Parse-SDK-Flutter#350]
This issue is part of #194 

**Should we add registerSubclass to match the Android/IOS SDK?**

**Pros:**
- usage of custom ParseObject classes would be much simpler

**Cons:**
- huge change
- existing code might break by updating the SDK

**Comment by phillwiggins** [comment]
Can you explain the pros in a bit more depth?

**Comment by maaeps** [comment]
Often the user can provide a copy-object. This is used to create new objects based on a custom ParseObject class.
By holding this copy-objects in the background, we would be able to create every object based on the right sub class.
As far as I know, included objects in a query are always of type ParseObject. This would be avoidable, too.

In summary: the user would not have to care about providing such copy-objects.

**Comment by phillwiggins** [comment]
Apologies for the late reply. I still don't actually understand this. Do you have you any examples where this is done already? I understand the Parse Android SDK does but as far as I'm aware there's not too much benefit when using Flutter as we don't have reflection to iterate through class structures?

**Comment by maaeps** [comment]
I did was not working on this for a while, too.
But I've started implementing this feature now.
To get an idea what I am talking about, you can checkout my [branch](https://github.com/maaeps/Parse-SDK-Flutter/tree/registerSubclass).
I you run something like this after the `Parse().initialize();`
```
ParseCoreData.instance
      ..registerSubClass("Test", () => Test1())
      ..registerSubClass("Test2", () => Test2())
      ..registerSubClass("Test3", () => Test3());
```

The SDK will use your custom classes for objects with the classnames 'Test', 'Test2' and 'Test3'.

This should allow you to do something like this:
`Test2 test = object.get<Test2>('test');`
Without this changes a error gets thrown: `type 'ParseObject' is not a subtype of type 'Test2'`

**Comment by maaeps** [comment]
...adding a Constructor to QueryBuilder where the object the registered sub-classes get uses automatically would be useful, too.

**Comment by phillwiggins** [comment]
I really like this idea. Seems very logical. Did you have any luck
implementing it?

**Comment by maaeps** [comment]
Well, I think something as simple as [this](https://github.com/maaeps/Parse-SDK-Flutter/commit/f157becbdc042bcb76b51c9ea6462d61c994c56c) should work.

**Comment by maaeps** [comment]
@phillwiggins 
I've added some more changes to [my branch](https://github.com/maaeps/Parse-SDK-Flutter/tree/registerSubclass).
I think while working with registerSubclass, there will appear unexpected behavior (the subclass is not used), as I am sure, I have not found every critical initialization of a ParseObject jet.

Should I keep this changes in my fork for now and create a PR once I have managed to build an app using registerSubclass OR should I create a PR to 1.27 now/soon?
I'm fine with both options.
